### PR TITLE
Fixes #527 - Deploy failed, check the error message in the Console View

### LINKED ISSUE #527: Deploy failed, check the error message in the Console View
We should put the error message right into the dialog we pop up on deploy failure; not simply say check the console view. 

image

### PR DESCRIPTION
**Pull request by akerekes**

### CODE REVIEW

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/ConsoleOutputBasedDeployErrorMessageProvider.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    this.outputLineListener = outputLineListener;
+  }
+
+  @Override
+  public String getErrorMessage() {
+    List<String> messages = outputLineListener.getCollectedMessages();
+    if (!messages.isEmpty()) {
+      return Joiner.on('\n').join(messages);
+    } else {
+      return null;
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/ConsoleOutputBasedDeployErrorMessageProvider.java]: would it make sense to return an empty list here?

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/messages.properties [hunk 2]
```diff
@@ -9,5 +9,7 @@ project.id.missing=App Engine project ID is not set, deploy failed.
 deploy.standard.runnable.name=Deploying to App Engine Standard
 deploy.job.stagingdir.create.failed=Cannot create staging directory.
 deploy.job.sourcedir.missing=Staging source directory does not exist.
+deploy.job.staging.failed=Staging failed, check the error message in the Console View
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/messages.properties]: failed. Check

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/messages.properties [hunk 3]
```diff
@@ -9,5 +9,7 @@ project.id.missing=App Engine project ID is not set, deploy failed.
...
+deploy.job.deploy.failed=Deploy failed, check the error message in the Console View
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/messages.properties]: failed. Check

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/ConsoleOutputBasedDeployErrorMessageProvider.java (hunk 1 above)
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/ConsoleOutputBasedDeployErrorMessageProvider.java]: I'll change it to return empty String.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployJob.java [hunk 4]
```diff
@@ -112,12 +113,16 @@ public IStatus runInWorkspace(IProgressMonitor monitor) throws CoreException {
           return cloudSdkProcessStatus;
         }
         // temporary way of error handling, after #439 is fixed, it'll be cleaner
-        return StatusUtil.error(getClass(), "Staging failed, check the error message in the Console View");
+        String errorMessage =
+            getErrorMessageOrDefault(Messages.getString("deploy.job.staging.failed"));
+        return StatusUtil.error(getClass(), errorMessage);
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployJob.java]: Unfortunately, this doesn't seem to show the error message when staging fails:

selection_007

FYI, to reproduce staging failure, change `appengine-web.xml` to be a malformed XML.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployJobConfig.java [hunk 5]
```diff
@@ -31,6 +32,7 @@
   private ProcessOutputLineListener stdoutLineListener;
   private ProcessOutputLineListener stderrLineListener;
   private DefaultDeployConfiguration deployConfiguration;
+  private DeployErrorMessageProvider errorMessageProvider;
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployJobConfig.java]: This seems an odd field for a config object.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployJob.java [hunk 6]
```diff
@@ -146,6 +151,19 @@ protected void canceling() {
     super.canceling();
   }
 
+  /**
+   * @return the error message obtained from <code>config.getErrorMessageProvider()</code> or
+   * <code>defaultMessage</code>
+   */
+  private String getErrorMessageOrDefault(String defaultMessage) {
+    String errorMessage = config.getErrorMessageProvider().getErrorMessage();
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployJob.java]: Getting the error message via a config object doesn't feel right.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployJob.java (hunk 4 above)
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployJob.java]: Yeah, the problem is that the error message is not prefixed by the 'ERROR` string. I wonder if we should file bug against CloudSDK to standardize their error message format versus implementing a pattern matching against all formats we encounter. I guess it should be the former.

**Comment by akerekes** [comment]
@briandealwis yes, the intent of the config object was to avoid a 6 (now 7) argument constructor.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployJob.java (hunk 4 above)
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployJob.java]: I've filed an issue for the App Engine Tools project to fix this.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/META-INF/MANIFEST.MF [hunk 7]
```diff
@@ -10,7 +10,8 @@ Bundle-ActivationPolicy: lazy
 Require-Bundle: com.google.cloud.tools.appengine;bundle-version="0.2.5",
  org.eclipse.ui.workbench,
  org.eclipse.core.commands,
- com.google.cloud.tools.eclipse.appengine.deploy
+ com.google.cloud.tools.eclipse.appengine.deploy,
+ com.google.cloud.tools.eclipse.sdk
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/META-INF/MANIFEST.MF]: You can move this to `Import-Package`.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployJob.java (hunk 4 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployJob.java]: I wonder if that `/tmp/appcfg646xxxxx` file has useful information that should be somehow shown or made available directly.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/standard/StandardDeployCommandHandler.java [hunk 8]
```diff
@@ -142,16 +140,23 @@ public void done(IJobChangeEvent event) {
     consoleManager.showConsoleView(messageConsole);
   }
 
-  private String getConsoleName(String project) {
+  private static String getConsoleName(String project) {
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/standard/StandardDeployCommandHandler.java]: `projectName`?

File: plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/OutputCollectorOutputLineListener.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+ */
+public class OutputCollectorOutputLineListener implements ProcessOutputLineListener {
+
+  private final String prefix;
+  private final List<String> collectedMessages = new ArrayList<>();
+  private final ProcessOutputLineListener wrappedListener;
+
+  /**
+   * @param prefix collects all lines starting with this prefix. If <code>prefix</code> is
+   * <code>null</code> or empty string, it will not collect any lines
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/OutputCollectorOutputLineListener.java]: This seems an odd choice?  Wouldn't null make sense to collect all lines?  Perhaps a better option would be to provide a `Predicate<String>`?

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/standard/StandardDeployCommandHandler.java [hunk 10]
```diff
@@ -122,7 +120,7 @@ private void launchDeployJob(IProject project, Credential credential)
                                               new DeployConsole.Factory());
 
     MessageConsoleStream outputStream = messageConsole.newMessageStream();
-    StandardDeployJobConfig config = getDeployJobConfig(project, credential,
+    StandardDeployJobConfig config = createDeployJobConfig(project, credential,
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/standard/StandardDeployCommandHandler.java]: Do you think having a separate method is clearer than inlining the builder here directly?

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/standard/StandardDeployCommandHandler.java [hunk 11]
```diff
@@ -142,16 +140,23 @@ public void done(IJobChangeEvent event) {
...
+                                nowString);
   }
 
-  private StandardDeployJobConfig getDeployJobConfig(IProject project, Credential credential,
-      IPath workDirectory, MessageConsoleStream outputStream,
-      DefaultDeployConfiguration deployConfiguration) {
+  private static StandardDeployJobConfig createDeployJobConfig(IProject project,
+                                                               Credential credential,
+                                                               IPath workDirectory,
+                                                               MessageConsoleStream outputStream,
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/standard/StandardDeployCommandHandler.java]: So I believe you should probably create two streams, one for stdout and the other for stderr (with a different colour).  You can see the `org.eclipse.debug.ui.console.ConsoleColorProvider` for an example of setting colours. (We should punt this to a different issue.)

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/standard/StandardDeployCommandHandler.java (hunk 10 above)
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/standard/StandardDeployCommandHandler.java]: Do you mean a method on the `StandardDeployJobConfig` class? Or a separate builder class with appropriate methods?

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/standard/StandardDeployCommandHandler.java (hunk 8 above)
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/standard/StandardDeployCommandHandler.java]: Fixed. It's the (GCP) projectId.

File: plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/OutputCollectorOutputLineListener.java (hunk 9 above)
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/OutputCollectorOutputLineListener.java]: yeah, good idea!

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/standard/StandardDeployCommandHandler.java (hunk 11 above)
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/standard/StandardDeployCommandHandler.java]: #1248

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployJob.java (hunk 4 above)
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployJob.java]: #1249

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/standard/StandardDeployCommandHandler.java (hunk 10 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/standard/StandardDeployCommandHandler.java]: I meant inlining `createDeployJobConfig()` here.  There are a lot of deploy-config objects being thrown around here and it's a bit confusing.  Would it make sense to rename this `StandardDeployJobConfig` as a _Parameterization_?  Or move this into the Job?

Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/standard/StandardDeployCommandHandler.java]: I just got rid of the config object, and added these objects to the job itself.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployJob.java [hunk 12]
```diff
@@ -165,7 +189,7 @@ private static void saveCredential(Path destination, Credential credential) thro
 
   private CloudSdk getCloudSdk(Path credentialFile) {
     CloudSdk cloudSdk = new CloudSdk.Builder()
-                          .addStdOutLineListener(config.getStdoutLineListener())
+                          .addStdOutLineListener(stdoutLineListener)
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployJob.java]: I believe you're missing `stderrLineListener`

File: plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/OutputCollectorOutputLineListener.java [hunk 13]
```diff
@@ -17,38 +17,42 @@
...
 import java.util.List;
 
 /**
- * A {@link ProcessOutputLineListener} which also collects output lines starting with a certain
- * prefix.
- * <p>
- * If the prefix is <code>null</code> or empty, it will not collect any lines.
+ * A {@link ProcessOutputLineListener} that collects output lines satisfying some {@link Predicate}.
  */
 public class OutputCollectorOutputLineListener implements ProcessOutputLineListener {
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/OutputCollectorOutputLineListener.java]: Strange to see _Output_ twice in the name?  `CollectingLineListener`?

File: plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/OutputCollectorOutputLineListener.java [hunk 14]
```diff
@@ -17,38 +17,42 @@
...
+  private final Predicate<String> condition;
   private final List<String> collectedMessages = new ArrayList<>();
   private final ProcessOutputLineListener wrappedListener;
 
   /**
-   * @param prefix collects all lines starting with this prefix. If <code>prefix</code> is
-   * <code>null</code> or empty string, it will not collect any lines
+   * @param predicate all lines satisfying this predicate will be collected
    */
   public OutputCollectorOutputLineListener(ProcessOutputLineListener wrappedListener,
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/OutputCollectorOutputLineListener.java]: Does it need to wrap another listener?  `CloudSdk` supports multiple listeners, right?  Seems a bit odd to have them all in one.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployJob.java [hunk 15]
```diff
@@ -68,37 +71,58 @@
...
+    this.project = project;
+    this.credential = credential;
+    this.workDirectoryParent = workDirectoryParent;
+    this.stdoutLineListener = stdoutLineListener;
+    this.stderrLineListener = stderrLineListener;
+    this.deployConfiguration = deployConfiguration;
     errorCollectingLineListener =
-        new OutputCollectorOutputLineListener(config.getStderrLineListener(),
-                                              ERROR_MESSAGE_PREFIX);
+        new OutputCollectorOutputLineListener(this.stderrLineListener,
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployJob.java]: Oh I see: here's the `stderrLineListener`.  I don't think this is right: you're now only passing matching lines which means anything written to stderr that doesn't match the predicate will be thrown away.

File: plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/OutputCollectorOutputLineListener.java (hunk 14 above)
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/OutputCollectorOutputLineListener.java]: Ah, that's a good point.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployJob.java (hunk 15 above)
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployJob.java]: Maybe the javadoc is misleading, but the collector will pass the line to the wrapped listener always. I'm thinking about the collector listener like a decorator.

File: plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/OutputCollectorOutputLineListener.java (hunk 13 above)
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/OutputCollectorOutputLineListener.java]: ok

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployJob.java (hunk 15 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployJob.java]: I misread: I thought the wrappedListener had been brought inside the `if`.  Sorry!

Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployJob.java]: But I guess it's still better not to wrap the listener, as https://github.com/GoogleCloudPlatform/google-cloud-eclipse/pull/1219#discussion_r96913253:

```java
    .addStdOutLineListener(config.getStdoutLineListener())
    .addStdErrLineListener(config.getStderrLineListener())
    .addStdErrLineListener(errorCollectingLineListener)
```

**Pull request merged**